vector/src/main/java/im/vector/app/features/onboarding/store/OnboardingStore.kt

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2021 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.features.onboarding.store
+
+import android.content.Context
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.core.edit
+import androidx.datastore.preferences.core.stringPreferencesKey
+import androidx.datastore.preferences.preferencesDataStore
+import im.vector.app.core.extensions.removeKeysWithPrefix
+import im.vector.app.features.onboarding.FtueUseCase
+import kotlinx.coroutines.flow.first
+
+private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "vector_onboarding")
+
+/**
+ * Local storage for:
+ * - messaging use case (Enum/String)
+ */
+class OnboardingStore constructor(
+        private val context: Context,
+        private val myUserId: String
+) {
+
+    suspend fun readUseCase() = context.dataStore.data.first().let { preferences ->
+        preferences[myUserId.toUseCaseKey()]?.let { FtueUseCase.from(it) }
+    }
+
+    suspend fun setUseCase(useCase: FtueUseCase) {
+        context.dataStore.edit { settings ->
+            settings[myUserId.toUseCaseKey()] = useCase.persistableValue
+        }
+    }
+
+    suspend fun resetUseCase() {
+        context.dataStore.edit { settings ->
+            settings.remove(myUserId.toUseCaseKey())
+        }
+    }
+
+    suspend fun clear() {
+        context.dataStore.removeKeysWithPrefix(myUserId)
+    }
+
+    private fun String.toUseCaseKey() = stringPreferencesKey("$this-use_case")
+}
```

Comment:
This class could be simplified with:
```kotlin
class OnboardingStore(
        private val context: Context,
        myUserId: String
) {
    private val useCasePreferences = stringPreferencesKey("$myUserId-use_case")

    suspend fun readUseCase() = context.dataStore.data.first().let { preferences ->
        preferences[useCasePreferences]?.let { FtueUseCase.from(it) }
    }

    suspend fun setUseCase(useCase: FtueUseCase) {
        context.dataStore.edit { settings ->
            settings[useCasePreferences] = useCase.persistableValue
        }
    }

    suspend fun resetUseCase() {
        context.dataStore.edit { settings ->
            settings.remove(useCasePreferences)
        }
    }

    suspend fun clear() {
        resetUseCase()
    }
}
```

and `removeKeysWithPrefix` is not necessary since we know have a new per userId store object.

I am also wondering if we could make it more generic, so its goal will be to store any data app side related to the session. I have in mind the storage of a client secret for https://github.com/matrix-org/sygnal/issues/70.

The class name could be something like `SessionDataStore` (open for better ideas!)

Reply:
sounds good to me :+1:

Reply:
went with `VectorSessionStore`
https://github.com/vector-im/element-android/pull/5009/commits/42f5ba8746c231f68ce7806bb7812a317edf133b